core-common/src/main/java/org/glassfish/jersey/internal/OsgiRegistry.java

```diff
@@ -552,14 +552,18 @@ private void register(final Bundle bundle) {
         }
     }
 
-    private List<Class<?>> locateAllProviders(final String serviceName) {
+    private List<Class<?>> locateAllProviders(final Class<?> serviceClass) {
         lock.readLock().lock();
         try {
             final List<Class<?>> result = new LinkedList<Class<?>>();
             for (final Map<String, Callable<List<Class<?>>>> value : factories.values()) {
-                if (value.containsKey(serviceName)) {
+                if (value.containsKey(serviceClass.getName())) {
                     try {
-                        result.addAll(value.get(serviceName).call());
+                        for (final Class<?> clazz : value.get(serviceClass.getName()).call()) {
+                            if (serviceClass.isAssignableFrom(clazz)) {
+                                result.add(clazz);
+                            }
+                        }
```

Comment:
is it possible to warn about the classes that are not loaded? Something like:
```
else {
   CodeSource src = serviceClass.getProtectionDomain().getCodeSource();
   if (src != null) {
       Logger.log(Level.FINE, "class {0} was not registered to OSGi", src.getLocation() + "!" + serviceClass.getName());
   } else {
       Logger.log(Level.FINE, "class {0} was not registered to OSGi", serviceClass.getName());
   }
}
```

Reply:
No problem. I was thinking about that, but was not sure how often the `ServiceFinder.find` is used so did not want to spam logs.
And I guess I will use lower debug level (FINER).

Reply:
Had to get rid of the `Class.getModule()` from the log message since it is not available in Java 8. But now, to get more info for the message, one would have to get it from the factory (`BundleSpiProvidersLoader`) bundle which I am not sure is worth it.

Reply:
The information from `CodeSource` would be nice, but I admit that OSGi env could have `CodeSource` null.